### ISSUE TITLE
Failed to Find Class 500 Error 

### ISSUE THREAD

**Issue by chenchik** [jcustenborder/kafka-connect-spooldir#26]
I am trying to add this plugin connector to my current instance of kafka connect. I'm using Openshift and am running Kafka Connect using the docker image below (it's basically an older version of confluentinc/cp-docker-images except it has the jar files from the previously mentioned GitHub and the hdfs connector jar is upgraded) :

https://hub.docker.com/r/chenchik/custom-connect-hdfs/

I run kafka connect with a huge command that sets a ton of environment variables. From what I understand the plugins should go in /etc/kafka-connect/jars, once they're in there, they should work. 

In order to install this plugin into my Kafka Connect instance I cloned this plugin from github and ran:

`mvn clean package`

Then I took all the files in /target and copied them into my container running Kafka Connect into the `/etc/kafka-connect/jars` directory. I didn't change any environment variables after that.

When I try to activate the connector using the REST API, I issue a POST request with this payload:

```
{
    "name": "csv-json-1",
    "config": {
        "connector.class": "com.github.jcustenborder.kafka.connect.spooldir.SpoolDirCsvSourceConnector",
        "tasks.max": "1",
        "finished.path":"/csv-json/results/",
	"input.file.pattern":".*csv",
	"error.path":"/csv-json/errors/",
	"topic":"danila-csv-json",
	"input.path":"/csv-json/input/",
	"key.schema":"com.github.jcustenborder.kafka.connect.spooldir.CsvSchemaGenerator",
	"value.schema":"com.github.jcustenborder.kafka.connect.spooldir.CsvSchemaGenerator"
    }
}
```

The response I get every time is:

{
    "error_code": 500,
    "message": "Failed to find any class that implements Connector and which name matches com.github.jcustenborder.kafka.connect.spooldir.SpoolDirCsvSourceConnector, available connectors are: org.apache.kafka.connect.source.SourceConnector, org.apache.kafka.connect.tools.MockSourceConnector, org.apache.kafka.connect.file.FileStreamSinkConnector, io.confluent.connect.hdfs.tools.SchemaSourceConnector, org.apache.kafka.connect.tools.VerifiableSourceConnector, io.confluent.connect.s3.S3SinkConnector, org.apache.kafka.connect.file.FileStreamSourceConnector, org.apache.kafka.connect.tools.VerifiableSinkConnector, io.confluent.connect.jdbc.JdbcSinkConnector, io.confluent.connect.jdbc.JdbcSourceConnector, io.confluent.connect.elasticsearch.ElasticsearchSinkConnector, org.apache.kafka.connect.sink.SinkConnector, io.confluent.connect.storage.tools.SchemaSourceConnector, org.apache.kafka.connect.tools.MockConnector, org.apache.kafka.connect.tools.MockSinkConnector, org.apache.kafka.connect.tools.SchemaSourceConnector, io.confluent.connect.hdfs.HdfsSinkConnector"
}

If I issue a GET request to /connector-plugins, it is not listed.

I also cannot seem to find any logs inside of the container that can explain what's going on. The only kind log message I get is from the log the container is providing openshift. This is the only entry that pops up:

[2017-08-01 22:22:35,635] INFO 172.17.0.1 - - [01/Aug/2017:22:22:15 +0000] "POST /connectors HTTP/1.1" 500 1081 20544 (org.apache.kafka.connect.runtime.rest.RestServer)

Any idea on what I can do to resolve this issue?

**Comment by jcustenborder** [comment]
@chenchik That basically happens when the connector is not in the classpath. I'd double check that the add command on your dockerfile isn't putting it in a subdirectory or something.

**Comment by MohammadChalaki** [comment]
I've been trying to run the csv-connect not using docker, so I've started zookeeper and kafka on a kafka cluster with three zookeepers and brokers.
Starting connect, I've used a distributed mode with a properties file like below:
myworker.properties:
[myworker.txt](https://github.com/jcustenborder/kafka-connect-spooldir/files/4769144/myworker.txt)

My plugin path directory containing dependencies is filled with jar files mentioned below:
animal-sniffer-annotations-1.17.jar 
commons-logging-1.2.jar
guava-27.1-jre.jar
kafka-connect-spooldir-1.0.31-sources.jar
checker-qual-2.5.2.jar
commons-text-1.3.jar   
j2objc-annotations-1.1.jar
kafka-connect-spooldir-2.0-SNAPSHOT.jar
commons-beanutils-1.9.3.jar
connect-utils-0.4.162.jar
jackson-annotations-2.9.0.jar  
listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
commons-collections-3.2.2.jar
error_prone_annotations-2.2.0.jar
jackson-core-2.9.8.jar
opencsv-4.5.jar
commons-collections4-4.2.jar
extended-log-format-0.0.2.13.jar
jackson-databind-2.9.8.jar
reflections-0.9.11.jar
commons-compress-1.20.jar
failureaccess-1.0.1.jar
javassist-3.21.0-GA.jar
value-2.5.5.jar
commons-lang3-3.8.1.jar
freemarker-2.3.28.jar
jsr305-3.0.2.jar


Then parsing a json to my connect using command:
curl -X POST -H "Content-Type: application/json" -d @sample.json http://localhost:8083/connectors | jq
which the json file is:
[sample.txt](https://github.com/jcustenborder/kafka-connect-spooldir/files/4769148/sample.txt)

but I get the error:
 "error_code": 500,
  "message": "Failed to find any class that implements Connector and which name matches com.github.jcustenborder.kafka.connect.spooldir.SpoolDirCsvSourceConnector

**Comment by jcustenborder** [comment]
@MohammadChalaki I would not use a path based on `~`. Fully qualify the path. Change

`plugin.path=~/spooldir/target/kafka-connect-target/usr/share/kafka-connect/kafka-connect-spooldir/`

to 

`/home/whoever/spooldir/target/kafka-connect-target/usr/share/kafka-connect/kafka-connect-spooldir/`

You can also download it prebuilt from [here](https://www.confluent.io/hub/jcustenborder/kafka-connect-spooldir)

**Comment by MohammadChalaki** [comment]
Unfortunately by changing ~/ to /home/ubuntu, as you've mentioned, my connect can't ba able to be up. I've also used prebuilt and mvn clean package to get dependencies.

**Comment by MohammadChalaki** [comment]
Unfortunately changing ~/... to home/ubnuntu/... didn't solve my problem.
I've also used the prebuilt and 'mvn clean package' to get all the dependencies in the specified plugin.path.

**Comment by jcustenborder** [comment]
@MohammadChalaki That sounds like a permissions problem. It's going to depend on how you are starting connect. For example if you're using systemd it's mostly likely using a different user that doesn't have access. You should try putting something like `plugin.path=/opt/kafka-connect`. Place the connector in `/opt/kafka-connect/kafka-connect-spooldir` and restart your worker. This will load each subdirectory under `/opt/kafka-connect` meaning you could add another connector and just restart the worker. If the permissions are working properly you should see the logs mentioning loading the connector.

**Comment by MohammadChalaki** [comment]
Sorry for the bother, but unfortunately I didn't get the  "Place the connector in" part. Could you please explain it more?!
Thnx!

**Comment by jcustenborder** [comment]
@MohammadChalaki All good buddy. When you run `mvn clean package` one of the outputs is a folder called `target/kafka-connect-target` under this it has all of the dependencies needed to run the connector. You need to take the content of this folder and copy it to `/opt/kafka-connect/kafka-connect-spooldir`. If your worker properties file has `plugin.path=/opt/kafka-connect` the worker will look at each subfolder in this folder and load them as a plugin. This requires a restart. Another way to do this is to install via the [Confluent Hub](https://www.confluent.io/hub/jcustenborder/kafka-connect-spooldir) this will do it for you. I do this a lot when I build docker containers for Kafka Connect. Check this [Dockerfile](https://github.com/jcustenborder/kafka-connect-all/blob/master/Dockerfile) out. Docker isn't required for connect FYI. This is just an example of how to use the Confluent hub.

**Comment by SachinKSunny** [comment]
Hey @jcustenborder, I used confluent-hub install jcustenborder/kafka-connect-spooldir:latest and I'm facing the error code 500. Can you help me understand why this might be happening? I installed JDBCConnector similarly and it shows up when I send a GET request to localhost:8083/connector-plugins but not the SpoolDirConnector.

**Comment by jcustenborder** [comment]
@SachinKSunny You need to make sure that all of the workers in your connect cluster get the plugin installed. You need to make sure that you restart all of the workers after installing a plugin. Verify that the plugin.path of your config matches where the connectors are installed. For example when I do this in docker I set my [plugin.path](https://github.com/jcustenborder/kafka-connect-all/blob/master/Dockerfile#L2) to include where the plugins are installed.

**Comment by SachinKSunny** [comment]
After much digging, I figured out what the issue was. I was setting the JAVA_HOME variable to point to JAVA_8(To fix some errors with Pentaho Data Integration) in my bash_profile. Once I removed that and was using the default JDK11, I was able to see the Spooldir connector plugins.

Leaving this comment here in case someone else runs into a similar issue or is a complete noob like me.

@jcustenborder: Thanks for the super-fast response. Maybe just reaching out to you was the luck I needed to figure the issue out. I was very close to pulling all of my hair out. Is your plugin built on JDK 11?
